core/src/main/scala/kafka/server/AdminManager.scala

```diff
@@ -0,0 +1,103 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package kafka.server
+
+import java.util.Properties
+
+import kafka.admin.AdminUtils
+import kafka.controller.KafkaController
+import kafka.log.LogConfig
+import kafka.metrics.KafkaMetricsGroup
+import kafka.utils._
+import org.apache.kafka.common.errors.InvalidRequestException
+import org.apache.kafka.common.metrics.Metrics
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.requests.CreateTopicsRequest._
+
+import scala.collection._
+import scala.collection.JavaConverters._
+
+class AdminManager(val config: KafkaConfig,
+                   val metrics: Metrics,
+                   val controller: KafkaController,
+                   val metadataCache: MetadataCache,
+                   val zkUtils: ZkUtils) extends Logging with KafkaMetricsGroup {
+  this.logIdent = "[Admin Manager on Broker " + config.brokerId + "]: "
+
+  /**
+    * Create topics and wait until the topics have been completely created.
+    * The callback function will be triggered either when timeout, error or the topics are created.
+    */
+  def createTopics(timeout: Int,
+                   createInfo: Map[String, TopicDetails],
+                   responseCallback: Map[String, Errors] => Unit) {
+
+    // 1. map over topics creating assignment and calling zookeeper
+    val brokers = metadataCache.getAliveBrokers.map { b => kafka.admin.BrokerMetadata(b.id, b.rack) }
+    val metadata = createInfo.map { case (topic, arguments) =>
+      try {
+        val configs = new Properties()
+        arguments.configs.asScala.foreach { case (key, value) =>
+          configs.setProperty(key, value)
+        }
+        LogConfig.validate(configs)
+
+        val assignments = {
+          if ((arguments.numPartitions != NO_NUM_PARTITIONS_SIGN || arguments.replicationFactor != NO_REPLICATION_FACTOR_SIGN)
```

Comment:
This came up as part of the "Client Side Auto Topic Creation" discussion. 

I am not in favor of adding the functionality as I believe client defaults will be more effective and flexible. And beyond that I think that a wider broker min/max configuration restriction discussion could be useful for admins. 

@ijuma seamed on the fence. He said:

> 2. Like Grant, I'm unsure that will generally be used in a positive way.
However, if this is what we need to be able to deprecate server auto-topic
creation, the benefits outweigh the costs in my opinion.

And @guozhangwang said:

> I agree with Grant and Ismael regarding 1 / 2 / 3, and for 2) I would
prefer having the default configs on the client side (i.e. like what we did
in the kafka-topics script today) than on the cluster side, where the
AdminClient will auto-set the configs if they are not specified by the user
triggering its function...

We can always add the functionality later if we decide its needed, otherwise client side defaults is the way forward.